flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushPartitionIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,325 @@
+			.statistic(tableSourceTable.getStatistic())
+			.tableStats(newTableStat)
+			.build();
+
+		String extraDigest = remainingPartitions == null ? "partitions=[]" :
+			("partitions=[" +
+				String.join(", ", remainingPartitions
+					.stream()
+					.map(Object::toString)
+					.toArray(String[]::new)) +
+				"]");
+		TableSourceTable newTableSourceTable = tableSourceTable.copy(dynamicTableSource, newStatistic, new String[]{extraDigest});
+
+		LogicalTableScan newScan = LogicalTableScan.create(scan.getCluster(), newTableSourceTable, scan.getHints());
+
+		RexNode nonPartitionPredicate = RexUtil.composeConjunction(rexBuilder, JavaConversions.seqAsJavaList(allPredicates._2()));
+		if (nonPartitionPredicate.isAlwaysTrue()) {
+			call.transformTo(newScan);
+		} else {
+			Filter newFilter = filter.copy(filter.getTraitSet(), newScan, nonPartitionPredicate);
+			call.transformTo(newFilter);
+		}
+	}
+
+	/**
+	 * adjust the partition field reference index to evaluate the partition values.
+	 * e.g. the original input fields is: a, b, c, p, and p is partition field. the partition values
+	 * are: List(Map("p"->"1"), Map("p" -> "2"), Map("p" -> "3")). If the original partition
+	 * predicate is $3 > 1. after adjusting, the new predicate is ($0 > 1).
+	 * and use ($0 > 1) to evaluate partition values (row(1), row(2), row(3)).
+	 */
+	private RexNode adjustPartitionPredicate(List<String> inputFieldNames, List<String> partitionFieldNames, RexNode partitionPredicate) {
+		return partitionPredicate.accept(new RexShuttle(){
+			@Override
+			public RexNode visitInputRef(RexInputRef inputRef) {
+				int index = inputRef.getIndex();
+				String fieldName = inputFieldNames.get(index);
+				int newIndex = partitionFieldNames.indexOf(fieldName);
+				if (newIndex < 0) {
+					throw new TableException(String.format("Field name '%s' isn't found in partitioned columns." +
+						" Validator should have checked that.", fieldName));
+				}
+				if (newIndex == index){
+					return inputRef;
+				} else {
+					return new RexInputRef(newIndex, inputRef.getType());
+				}
+			}
+		});
+	}
+
+	private Optional<List<Map<String, String>>> readPartitionFromCatalogAndPrune(
+			FlinkContext context,
+			Catalog catalog,
+			ObjectIdentifier tableIdentifier,
+			List<String> allFieldNames,
+			Seq<RexNode> partitionPredicate,
+			Function<List<Map<String, String>>, List<Map<String, String>>> pruner) throws TableNotExistException, TableNotPartitionedException{
+		RexNodeToExpressionConverter converter = new RexNodeToExpressionConverter(
+			allFieldNames.toArray(new String[0]),
+			context.getFunctionCatalog(),
+			context.getCatalogManager(),
+			TimeZone.getTimeZone(context.getTableConfig().getLocalTimeZone()));
+		ArrayList<Expression> partitionFilters = new ArrayList<>();
+		Option<ResolvedExpression> subExpr;
+		for (RexNode node: JavaConversions.seqAsJavaList(partitionPredicate)) {
+			subExpr = node.accept(converter);
+			if (!subExpr.isEmpty()) {
```

Comment:
what if the `subExpr` is empty ?

Reply:
we only use `subExpr` to accept value that is calculated from node. we have checked whether `subExpr` isEmpty in line 276

Reply:
if `subExpr` is empty, its corresponding sub-filter is dropped, then the result is incorrect. There is no test coverage...